### PR TITLE
Revise prelu

### PR DESCRIPTION
**Pull request by pszmel**
### Details:
 - update prelu ref impl
 - add backend tests
 - removed few slope shapes from from SLT, because they were non broadcastable
 - add visitors tests
 
### Tickets:
 - 37461

### CODE REVIEW

File: ngraph/core/reference/include/ngraph/runtime/reference/prelu.hpp [hunk 1]
```diff
@@ -24,11 +25,37 @@ namespace ngraph
                        const Shape& arg_shape,
                        const Shape& slope_shape)
             {
+                if (arg_shape.size() >= 2)
+                {
+                    NGRAPH_CHECK(arg_shape[1] == shape_size(slope_shape), "If input shape is >= 2, then second second dim of input needs to be equal to slope shape");
```
Comment (mitruska) [inline ngraph/core/reference/include/ngraph/runtime/reference/prelu.hpp:30]: [Some onnx tests](https://dev.azure.com/openvinoci/dldt/_build/results?buildId=71942&view=logs&j=0609cd82-f9f7-5e70-371c-89fca78031ab&t=d7ecf6fe-a4b9-53ec-d3c3-e8ce906215b2&l=615) fail with this PR changes but [pass ](https://dev.azure.com/openvinoci/dldt/_build/results?buildId=71942&view=logs&j=0609cd82-f9f7-5e70-371c-89fca78031ab&t=d7ecf6fe-a4b9-53ec-d3c3-e8ce906215b2&l=3111)with current master reference and CPU implementation. 
So it is obvious regression.

**Review by mitruska — COMMENTED** [review]
The [PRelu](https://github.com/openvinotoolkit/openvino/blob/master/docs/ops/activation/PReLU_1.md) spec was updated with restriction to 1D tensor for `slope` input, that was discussed in https://github.com/openvinotoolkit/openvino/pull/5016#discussion_r603877858 but it breaks compatibility with [ONNX Prelu](https://github.com/onnx/onnx/blob/master/docs/Operators.md#PRelu) which allows already broadcasted second input ([failing onnx prelu test](https://dev.azure.com/openvinoci/dldt/_build/results?buildId=71942&view=logs&j=0609cd82-f9f7-5e70-371c-89fca78031ab&t=d7ecf6fe-a4b9-53ec-d3c3-e8ce906215b2&l=615) -> [prelu onnx model](https://github.com/openvinotoolkit/openvino/blob/78076656fd1759cf449dc3ae7d4ccc2927e07ae9/ngraph/test/models/onnx/prelu.prototxt#L30-L45)). @ggalieroc @lazarevevgeny

**Comment by IvanNovoselov** [comment]
An analysis of conformance tests suggests that the reason why most of the PRelu tests fail on CPU may be connected to the reference implementation. Could you please check two cases for the revised implementation:
1)	Input data: {-10, -10, -10, -10}, Input shape: {1, 2, 1, 2}, Slope input: {0.1, 10} => Produced output: {-1, -1, -100, -100}, not {-1, -100, -1, -100}.
2)	Input data: {-1, 0, -1, -1, -1}, Input shape: {1, 5, 1, 1}, Slope input: {1, 2, 3, 4, 5} => Produced output: {-1, 0, -3, -4, -5}, not {-1,  0, -2, -3, -4}.

**Comment by pszmel** [comment]
> ts suggests that the reason why most of the PRelu tests fail on CPU may be connected to the reference implementation. Could you pleas

@IvanNovoselov In the first test I set up slope shape as {2} and the output is {-1, -100, -1, -100}. In the second test if I set up slope shape as {5} then there is a error both on CPU and INTERPRETER caused by invalid slope shape, but if I change the slope shape to {1, 5, 1, 1}, then the output is {-1, 0, -3, -4, -5}

**Comment by IvanNovoselov** [comment]
> > ts suggests that the reason why most of the PRelu tests fail on CPU may be connected to the reference implementation. Could you pleas
> 
> @IvanNovoselov In the first test I set up slope shape as {2} and the output is {-1, -100, -1, -100}. In the second test if I set up slope shape as {5} then there is a error both on CPU and INTERPRETER caused by invalid slope shape, but if I change the slope shape to {1, 5, 1, 1}, then the output is {-1, 0, -3, -4, -5}

@pszmel Thank you for the tests! Results of the second test seems to be fine, but are you sure that the Op performed as expected in the first test?

I'm concerned because PReLU is expected to fold slopes with channels according to the specs (please see the [text](https://github.com/openvinotoolkit/openvino/blame/1e3b06f439bd9e182461dc860abc3851a57f693a/docs/ops/activation/PReLU_1.md#L22) or [the last example](https://github.com/openvinotoolkit/openvino/blame/1e3b06f439bd9e182461dc860abc3851a57f693a/docs/ops/activation/PReLU_1.md#L91-L115)). It seems however that the slopes are folded with the last dimension, which is width by default.
Here I presume that the default layout is NCHW, it means that the channel dimension is 1 and the number of channels = input_shape[1] = 2. Likewise, the width dimension is the last one (4), and the data width is  input_shape[3] = 2.

Keeping in mind this ambiguity, it might be a good idea to explicitly specify the default layout in the specs.

**Comment by IvanNovoselov** [comment]
> > > > ts suggests that the reason why most of the PRelu tests fail on CPU may be connected to the reference implementation. Could you pleas
> > > 
> > > 
> > > @IvanNovoselov In the first test I set up slope shape as {2} and the output is {-1, -100, -1, -100}. In the second test if I set up slope shape as {5} then there is a error both on CPU and INTERPRETER caused by invalid slope shape, but if I change the slope shape to {1, 5, 1, 1}, then the output is {-1, 0, -3, -4, -5}
> > 
> > 
> > @pszmel Thank you for the tests! Results of the second test seems to be fine, but are you sure that the Op performed as expected in the first test?
> > I'm concerned because PReLU is expected to fold slopes with channels according to the specs (please see the [text](https://github.com/openvinotoolkit/openvino/blame/1e3b06f439bd9e182461dc860abc3851a57f693a/docs/ops/activation/PReLU_1.md#L22) or [the last example](https://github.com/openvinotoolkit/openvino/blame/1e3b06f439bd9e182461dc860abc3851a57f693a/docs/ops/activation/PReLU_1.md#L91-L115)). It seems however that the slopes are folded with the last dimension, which is width by default.
> > Here I presume that the default layout is NCHW, it means that the channel dimension is 1 and the number of channels = input_shape[1] = 2. Likewise, the width dimension is the last one (4), and the data width is input_shape[3] = 2.
> > Keeping in mind this ambiguity, it might be a good idea to explicitly specify the default layout in the specs.
> > Edit: It is explicitly stated in the specs that channels is the second dimension, sorry I didn't notice. Ok, it means that the first test should produce {-1, -1, -100, -100}.
> 
> @IvanNovoselov Actually both CPU and INTERPRETER produce {-1, -100, -1, -100}, because after broadcasting slope we get {0.1, 10, 0.1, 10}. Is it not the expected output?

@pszmel, current specs state that the slopes are given per-channel. The first two outputs correspond to the first channel, so I would expect them to use the first slope. Likewise with the second two outputs. The output is therefore expected to be {-1, -1, -100, -100}. 

Unfortunately, I could not reproduce your results regarding CPU plugin output. Current master version yields {-1, -1, -100, -100} for this test according to my analysis. Could you please add this test to the [activation.cpp](https://github.com/pszmel/openvino/blob/4b2801b904b9eebf1a1f36f5389fde654c56e1ee/inference-engine/tests/functional/plugin/cpu/shared_tests_instances/single_layer_tests/activation.cpp#L60-L78) to eliminate this concern?

It is also worth to add several similar tests, where the slope input consists of unique floats (not just all equal to -0.01), and slope broadcasting takes place. It is important, to use non-equal slopes because otherwise it is hard to spot inconsistencies in broadcasting rules between the reference and a plugin.

File: ngraph/test/backend/prelu.in.cpp [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+#include <algorithm>
+#include <cinttypes>
+#include <cmath>
+#include <cstdlib>
+#include <random>
+#include <string>
+#include "util/random.hpp"
+
+// clang-format off
+#ifdef ${BACKEND_NAME}_FLOAT_TOLERANCE_BITS
```
Comment (ilyachur) [inline ngraph/test/backend/prelu.in.cpp:14]: Why do we need to have these defines?

Comment (pszmel) [reply ngraph/test/backend/prelu.in.cpp]: My bad. Removed

**Comment by pszmel** [comment]
I am revoking to draft status, because there was a change in cpu plugin related to prelu

**Pull request closed without merging**